### PR TITLE
Remove `RequestLog` dependency in `RetryingClient`

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Related: #5708, #4834

If a decorator directly returns a response, the `RequestLog` for the request won't be complete. Because the `RequestLog` is automatically completed when it starts to be written to the wire by `AbstractRequestHandler`.

In the case of `RetryingClient`, `RequestLog`'s events are used to get exceptions to apply `RetryRule`. If the `RequestLog` events aren't complete, `RetryingClient` will fall into a deadlock.

Typically, `RequestLog` is automatically completed so it is difficult for the user to recognize that `RequestLog` must be completed. Consequently, it is easy to miss.

I propose two changes in this PR to solve the problem.

1. If `RequestLog.requestFirstBytesTransferredTimeNanos` isn't logged when a response completes, we can assume that the response is returned by a decorator. So a hook is added to a response completion event to complete the log manually for that case.

2. `RetryingClient` is refactored not to retry on `RequestLog` but to use `SplitHttpResponse` and subscribe to headers. An exception when headers are received is used to apply `RetryRule`.

Modifications:

- Add a hook that completes `RequestLog` if it is written to wire to `ClientUtil`
- Split `handleResponse` into `handleAggregatedResponse` and `handleStreamingResponse` in `RetryingClient`
  - Remove `RequestLog` usage in `handleStreamingResponse` and use `SplitHttpRequest` to check if an exception is raised in the response.

Result:

- Fixed deadlock that could occur in `RetryingClient` due to `RequestLog` not being completed.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/client/RuleFilter.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    private final BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter;
+    private final @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter;
+    private final @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter;
+    private final @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter;
+    private final @Nullable BiPredicate<ClientRequestContext, HttpHeaders> grpcTrailersFilter;
+    private final @Nullable BiPredicate<ClientRequestContext, Duration> totalDurationFilter;
+    private final boolean hasResponseFilter;
+
+    private RuleFilter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/client/RuleFilter.java:64]: `AbstractRuleBuilderUtil` is renamed to `RuleFilter` and is refactored to implement `BiFunction`.
```diff
diff --git Users/ikhoon/src/armeria/core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java Users/ikhoon/src/armeria/core/src/main/java/com/linecorp/armeria/internal/client/RuleFilter.java
index 1de4aceeb..e9c8a061b 100644
--- Users/ikhoon/src/armeria/core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java
+++ Users/ikhoon/src/armeria/core/src/main/java/com/linecorp/armeria/internal/client/RuleFilter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2025 LINE Corporation
+ * Copyright 2020 LINE Corporation
  *
  * LINE Corporation licenses this file to you under the Apache License,
  * version 2.0 (the "License"); you may not use this file except in compliance
@@ -23,6 +23,7 @@ import java.util.function.BiPredicate;
 import com.linecorp.armeria.client.AbstractRuleBuilder;
 import com.linecorp.armeria.client.ClientRequestContext;
 import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
 import com.linecorp.armeria.common.RequestHeaders;
 import com.linecorp.armeria.common.ResponseHeaders;
 import com.linecorp.armeria.common.annotation.Nullable;
@@ -31,16 +32,15 @@ import com.linecorp.armeria.common.logging.RequestLogProperty;
 import com.linecorp.armeria.common.util.Exceptions;
 import com.linecorp.armeria.internal.common.InternalGrpcWebTrailers;
 
-public final class AbstractRuleBuilderUtil {
+public final class RuleFilter implements BiFunction<ClientRequestContext, Throwable, Boolean> {
 
     /**
      * Merges the filters of {@link AbstractRuleBuilder} that returns {@code true}
      * if all rules in the {@link AbstractRuleBuilder} match
      * a given {@link ClientRequestContext} and {@link Throwable}.
      */
-    // TODO(ikhoon): Use BiPredicate.
-    public static BiFunction<? super ClientRequestContext, ? super Throwable, Boolean>
-    buildFilter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
+    public static RuleFilter of(
+            BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
             @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter,
             @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter,
             @Nullable BiPredicate<ClientRequestContext, HttpHeaders> grpcTrailersFilter,
@@ -48,14 +48,11 @@ public final class AbstractRuleBuilderUtil {
             @Nullable BiPredicate<ClientRequestContext, Duration> totalDurationFilter,
             boolean hasResponseFilter) {
 
-        return new Filter(requestHeadersFilter, exceptionFilter, responseHeadersFilter,
+        return new RuleFilter(requestHeadersFilter, exceptionFilter, responseHeadersFilter,
                               responseTrailersFilter, grpcTrailersFilter, totalDurationFilter,
                               hasResponseFilter);
     }
 
-    private AbstractRuleBuilderUtil() {}
-
-    private static class Filter implements BiFunction<ClientRequestContext, Throwable, Boolean> {
     private final BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter;
     private final @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter;
     private final @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter;
@@ -64,7 +61,7 @@ public final class AbstractRuleBuilderUtil {
     private final @Nullable BiPredicate<ClientRequestContext, Duration> totalDurationFilter;
     private final boolean hasResponseFilter;
 
-        Filter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
+    private RuleFilter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
                        @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter,
                        @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter,
                        @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter,
@@ -83,12 +80,10 @@ public final class AbstractRuleBuilderUtil {
     @Override
     public Boolean apply(ClientRequestContext ctx, Throwable cause) {
         final RequestLog log = ctx.log().partial();
-            if (log.isAvailable(RequestLogProperty.REQUEST_HEADERS)) {
-                final RequestHeaders requestHeaders = log.requestHeaders();
-                if (!requestHeadersFilter.test(ctx, requestHeaders)) {
+        final HttpRequest request = ctx.request();
+        if (request != null && !requestHeadersFilter.test(ctx, request.headers())) {
             return false;
         }
-            }
 
         // Safe to return true since no filters are set
         if (exceptionFilter == null && responseHeadersFilter == null &&
@@ -134,7 +129,6 @@ public final class AbstractRuleBuilderUtil {
             }
 
             // Suppressing the inspection rule because we don't want to return false too early.
-                //noinspection RedundantIfStatement
             if (maybeGrpcTrailers != null && grpcTrailersFilter.test(ctx, maybeGrpcTrailers)) {
                 // Found the matching gRPC trailers.
                 return true;
@@ -151,4 +145,3 @@ public final class AbstractRuleBuilderUtil {
         return false;
     }
 }
-}
```

**Comment by ikhoon** [comment]
This is ready for review.

**Review by jrhee17 — APPROVED** [review]
Understood the changes

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java [hunk 2]
```diff
@@ -177,6 +190,24 @@ O pushAndExecute(U delegate, ClientRequestContext ctx, I req) throws Exception {
         }
     }
 
+    private static <O extends Response> void completeLogIfIncomplete(ClientRequestContext ctx, O response) {
+        response.whenComplete().handle((unused, cause) -> {
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java:194]: If we need to support such cases, we may consider adding an option to defer `RequestLog` completion.

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java [hunk 3]
```diff
@@ -177,6 +190,24 @@ O pushAndExecute(U delegate, ClientRequestContext ctx, I req) throws Exception {
...
+            if (!logBuilder.isAvailable(RequestLogProperty.REQUEST_FIRST_BYTES_TRANSFERRED_TIME)) {
+                // As the request didn't reach AbstractHttpRequestHandler, the log won't be completed
+                // automatically. We need to end the request and response manually.
+                if (cause != null) {
+                    logBuilder.endRequest(cause);
+                    logBuilder.endResponse(cause);
+                } else {
+                    logBuilder.endRequest();
+                    logBuilder.endResponse();
+                }
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java]: Done. `ResponseCompletionException` is used to cancel `ctx` to successfully complete the log if `cause == null`.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 4]
```diff
@@ -364,77 +375,126 @@ private void handleResponseWithoutContent(RetryConfig<HttpResponse> config, Clie
...
+            logBuilder.endRequest();
+            logBuilder.responseHeaders(response.headers());
+            if (!response.trailers().isEmpty()) {
+                logBuilder.responseTrailers(response.trailers());
+            }
+            logBuilder.endResponse();
+        }
+    }
+
+    private static void completeLogIfIncomplete(HttpResponse response, ResponseHeaders headers,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:478]:
```suggestion
    private static void completeLogIfIncomplete(HttpResponse response, @Nullable ResponseHeaders headers,
```

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 5]
```diff
@@ -364,77 +375,126 @@ private void handleResponseWithoutContent(RetryConfig<HttpResponse> config, Clie
...
+
+    private static void completeLogIfIncomplete(AggregatedHttpResponse response, ClientRequestContext ctx) {
+        if (!ctx.log().isAvailable(RequestLogProperty.REQUEST_FIRST_BYTES_TRANSFERRED_TIME)) {
+            final RequestLogBuilder logBuilder = ctx.logBuilder();
+            logBuilder.endRequest();
+            logBuilder.responseHeaders(response.headers());
+            if (!response.trailers().isEmpty()) {
+                logBuilder.responseTrailers(response.trailers());
+            }
+            logBuilder.endResponse();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:474]: Shouldn't we also set the response content to null in case it's deferred?
https://github.com/line/armeria/blob/5d07d39ed516b32b6839bcc803578612efbb7a44/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java#L542-L548

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 6]
```diff
@@ -319,36 +322,44 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
...
-            // XXX(ikhoon): Should we use `response.aggregateWithPooledObjects()`?
             response.aggregate().handle((aggregated, cause) -> {
                 if (cause != null) {
+                    derivedCtx.logBuilder().endRequest(cause);
                     derivedCtx.logBuilder().endResponse(cause);
                     handleResponseWithoutContent(config, ctx, rootReqDuplicator, originalReq, returnedRes,
                                                  future, derivedCtx, HttpResponse.ofFailure(cause), cause);
-                    return null;
+                } else {
+                    completeLogIfIncomplete(aggregated, derivedCtx);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:340]: `completeLogIfBytesNotTransferred`?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 5 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:474]: Do we have to call `responseContent(null, null)` and `responseContentPreview(null)` when we end `RequestLog` manually? I'd like to clarify in which cases they should be set to null.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:475]: The `RequestLog` might not be complete because the properties are deferred when content previewing client is used?
https://github.com/line/armeria/blob/ce03ff4333de56c3facf515bf0587245ca36aa4d/core/src/main/java/com/linecorp/armeria/client/logging/ContentPreviewingClient.java#L160

File: core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java [hunk 7]
```diff
@@ -124,21 +124,14 @@ public final O execute(ClientRequestContext ctx, I req) throws Exception {
...
         return deferred;
     }
 
-    private static void completeExceptionally(ClientRequestContext ctx,
-                                              CompletableFuture<?> resFuture, Throwable t) {
-        resFuture.completeExceptionally(t);
-        ctx.logBuilder().endRequest(t);
-        ctx.logBuilder().endResponse(t);
-    }
-
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java:141]: 👍

**Review by trustin — APPROVED** [review]
LGTM! 👍 Please merge once @minwoox's comments are addressed.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 5 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:475]: Would it be simpler to fix `ContentPreviewingClient` to set `responseContentPreview(null)` if an exception is raised by the delegate? For example:
```java
try {
    ctx.logBuilder().defer(RequestLogProperty.RESPONSE_CONTENT_PREVIEW);
    final HttpResponse res = unwrap().execute(ctx, req);
    return setUpResponseContentPreviewer(contentPreviewerFactory, ctx, res, responsePreviewSanitizer);
} catch (Throwable e) {
    ctx.logBuilder().responseContentPreview(null);
    return HttpResponse.ofFailure(e);
}
```

It seems other cases are covered by 
https://github.com/line/armeria/blob/a5be69d99f03f892abfeca017597340e2b749eea/core/src/main/java/com/linecorp/armeria/internal/logging/ContentPreviewingUtil.java#L116-L132

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:475]: That probably doesn't work in this situation?
UserClient -> content previewing client -> retrying client -> short circuit client

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:475]: I added `ContentPreviewingClientLogCompletionTest` to check if child `RequestLog`s are complete. According to the test results, all logs are completed. 

If there’s a corner case not covered by that test, please let me know.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:475]: Thanks for adding the test. 😉

**Review by minwoox — APPROVED** [review]
👍 👍 👍
